drivers/display/display_st75256.c

```diff
@@ -0,0 +1,605 @@
+	err = st75256_write_command(dev, ST75256_SLEEP_OUT, NULL, 0);
+	if (err < 0) {
+		return err;
+	}
+	/* Wait 10 msec to allow display out of sleep */
+	k_msleep(10);
+	return st75256_write_command(dev, ST75256_DISPLAY_ON, NULL, 0);
+}
+
+static int st75256_set_window(const struct device *dev, int x, int y, int width, int height)
+{
+	struct st75256_data *data = dev->data;
+	int ret;
+	uint8_t x_position[] = {x, x + width - 1};
+	uint8_t y_position[2];
+
+	if (data->current_pixel_format == PIXEL_FORMAT_L_8) {
+		y_position[0] = y / 4;
+		y_position[1] = ((y + height) / 4) - 1;
+	} else {
+		y_position[0] = y / 8;
+		y_position[1] = ((y + height) / 8) - 1;
+	}
+
+	ret = st75256_write_command(dev, ST75256_EXTCOM_1, NULL, 0);
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = st75256_write_command(dev, ST75256_PAGE_RANGE, y_position, 2);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return st75256_write_command(dev, ST75256_COL_RANGE, x_position, 2);
+}
+
+static int st75256_start_write(const struct device *dev)
+{
+	int ret;
+
+	ret = st75256_write_command(dev, ST75256_EXTCOM_1, NULL, 0);
+	if (ret < 0) {
+		return ret;
+	}
+
+	return st75256_write_command(dev, ST75256_START_WRITE, NULL, 0);
+}
+
+static int st75256_write_pixels_MONO01(const struct device *dev, const uint16_t x, const uint16_t y,
+				       const uint8_t *buf,
+				       const struct display_buffer_descriptor *desc)
+{
+	const struct st75256_config *config = dev->config;
+	struct display_buffer_descriptor mipi_desc;
+	int ret;
+
+	for (int i = 0; i < desc->height / 8; i++) {
+		st75256_set_window(dev, x, y + i * 8, desc->width, desc->height);
+		st75256_start_write(dev);
+		mipi_desc.buf_size = desc->width;
+		mipi_desc.width = desc->width;
+		mipi_desc.height = 8;
+		ret = mipi_dbi_write_display(config->mipi_dev, &config->dbi_config, buf, &mipi_desc,
+					     PIXEL_FORMAT_MONO01);
+		if (ret < 0) {
+			return ret;
+		}
+	}
+	return mipi_dbi_release(config->mipi_dev, &config->dbi_config);
+}
+
+static int st75256_write_pixels_L_8(const struct device *dev, const uint16_t x, const uint16_t y,
+				    const uint8_t *buf,
+				    const struct display_buffer_descriptor *desc)
+{
+	const struct st75256_config *config = dev->config;
+	struct display_buffer_descriptor mipi_desc;
+	int32_t line_count = desc->height;
+	int ret, l;
+	int line_total = 0;
+
+	mipi_desc.pitch = desc->pitch;
+
+	st75256_set_window(dev, x, y, desc->width, desc->height);
+	st75256_start_write(dev);
+	while (line_count > line_total) {
+		l = 0;
+
+		/* Convert what the conversion buffer can hold to
+		 * pixelx[7:6], pixelx+pitch[5:4], pixelx+2pitch[3:2], pixelx+3pitch[1:0]
+		 */
+		for (; l * desc->width / 4 < config->conversion_buf_size &&
+		       line_count > line_total + l;
+		     l += 4) {
+			int i_lt = line_total + l;
+
+			for (int i = 0; i < desc->width &&
+					i + l * desc->width / 4 < config->conversion_buf_size;
+			     i++) {
+				int i_l = i + l * desc->width / 4;
+
+				config->conversion_buf[i_l] =
```

Comment:
Kind of unclear here- based on the documentation for the display, it seems like an 8 bit value represents 4 vertical pixels in grayscale mode, right? I guess in this mode we are packing the grayscale values from 4 vertical pixels?

I guess my question is why support this mode versus MONO? It seems like the additional conversion step will slow down refresh rates a lot, and we are dropping all luminance data. Not going to block this addition, just want to understand the reasoning.

Reply:
You are correct, it's 2 bits per pixels vtiled, should I add a comment about it?
It supports both mono and 4-level greyscale, I implemented both (st75256_write_pixels_MONO01 is also here and the driver defaults to mono), some people prefer having 4 levels of colors vs 2, I am one of those people.

Reply:
Yeah, I would prefer a comment just to be clear what is going on. Just interested, how are the performance implications of this? Also I see you have several different drivers where you are using conversion buffers to send data in the `L8` format, I guess the issue is that these display controllers lack a standardized format for their "multi bit grayscale" data?

Reply:
yea, the ssd13xx are all pretty similar and could be standardized, even arguably condensed into a base driver where just the registers and init change per driver, but st75256 and sh1122 are completely different.

Reply:
and for performance I dont really know, it should be pretty easy to optimize for the compiler because it's simple loops (and so likely can be auto-vectorized), but i havent really bothered to measure it and saw no dramatic difference on nrf52840

Reply:
Done.